devito/passes/iet/languages/C.py

```diff
@@ -1,16 +1,67 @@
 import numpy as np
 from sympy.printing.c import C99CodePrinter
 
-from devito.ir import Call, BasePrinter
+from devito.exceptions import InvalidOperator
+from devito.ir import Call, BasePrinter, List
 from devito.passes.iet.definitions import DataManager
 from devito.passes.iet.orchestration import Orchestrator
 from devito.passes.iet.langbase import LangBB
 from devito.symbolics import c_complex, c_double_complex
+from devito.symbolics.extended_sympy import UnaryOp
 from devito.tools import dtype_to_cstr
 
 __all__ = ['CBB', 'CDataManager', 'COrchestrator']
 
 
+class RealExt(UnaryOp):
+
+    _op = '__real__ '
+
+
+class ImagExt(UnaryOp):
+
+    _op = '__imag__ '
+
+
+def atomic_add(i, pragmas, split=False):
```

Comment:
with some gymnastics (e.g., adding a `cls` or `callback` parameter to construct the real and img parts), I think we could have just one generic `atomic_add`, like this one for example, and then reuse it in CXX too)